helix-term: introduce cdf / change-directory-to-current-file

cole-h opened the pull request:
Fixes: https://github.com/helix-editor/helix/issues/3519.

cole-h commented:
It's been a while since I last contributed here, so please don't be shy on telling me how I can improve / make this better :)

AceofSpades5757 commented:
IMHO
Due to how Helix likes to handle things, wouldn't it make more sense to go to the project root directory, like a "home" equivalent, with the fallback being the parent of the current file? There's another pr that's working on shell expansion that I think would solve this problem a little better.

cole-h commented:
I don't mind if there's a command for going to the project root, but I would _really_ like a command like this that allows me to go to the directory a file is in, and _not_ the project root. I usually start helix from the project root anyways, so that would be a completely useless no-op in almost every case of mine. 

Here's an example where I would find this supremely useful: in Nix, where relative paths are relative to the file referencing the path, it's nice to be able to make a new file just by `:cd` and then `:o newfile.nix` instead of having to type out the entire directory tree to where it should be located. `:o some/long/list/of/directories/before/we/get/to/where/we/need/to/place/the/newfile.nix`. Maybe slightly contrived, but hopefully communicates one of my usecases.

I'd be glad to introduce another command with this ability (`cd-to-file`, and `cdf`, maybe?) so that this other PR can do it's thing with `cd`, if that's deemed a better idea.

If there is no interest in this functionality whatsoever, please let me know and I'll move on.

poliorcetics commented:
I think `cd` alone should go to the "home" of the project for easier mental mapping: real `cd` goes to `$HOME`, helix's `cd` goes to project home.

Adding a `cdf` is a good idea though, I have often missed it. But maybe what you're missing is more of a new command to create a file next to another, like `ont`, `open-next-to`, where the default argument would be the current file but could be any file path

CBenoit commented:
> I think `cd` alone should go to the "home" of the project for easier mental mapping: real `cd` goes to `$HOME`, helix's `cd` goes to project home.

I agree with that, it feels more natural to me. I also expect this behavior to come in handy once a command such as `cd-to-file` is added.

> Adding a `cdf` is a good idea though, I have often missed it.

I think it would be a useful command too, but I’m a bit unsure about the naming. What about `cd-here`?

> But maybe what you're missing is more of a new command to create a file next to another, like `ont`, `open-next-to`, where the default argument would be the current file but could be any file path

I really like the `open-next-to` command idea.

cole-h commented:
I've rebased on the latest `master` and regenerated the docs. I reverted my changes to `cd` and instead introduce a new `cdf` / `change-directory-to-current-file` that implements my desired functionality. I'll leave the "`cd` without args should go to the project's root" functionality to another PR.

cole-h commented on helix-term/src/commands/typed.rs:
```diff
@@ -1897,6 +1925,13 @@ pub const TYPABLE_COMMAND_LIST: &[TypableCommand] = &[
+        TypableCommand {
+            name: "change-directory-to-current-file",
+            aliases: &["cdf"],
+            doc: "Change the current working directory to the parent of the current document.",
```
I wonder if this should become:
```suggestion
            doc: "Change the current working directory to the parent of the current file.",
```

or if everything else referencing "file" should become "document"...

the-mikedavis commented:
I think the change here looks good. I wonder if it's necessary to optimize this use-case so much though? There are a few other ways that I think could be fast enough:

* Having a register for the current filename, using `:cd`, inserting the filename, using `C-w` to delete back to the parent directory of that path, and then `<ret>`
* Having a variable for the current file's parent directory and `:cd $parent`
* Having a variable only for the current file and using `dirname` and some shell expansion to `cd` to that directory

cole-h commented:
Hi there! Now that 23.10 has released, is there anything I can do to help get this merged?

pascalkuthe commented:
> Having a register for the current filename, using :cd, inserting the filename, using C-w to delete back to the parent directory of that path, and then <ret>

this has been added in the latest release, with the command expansion PR we want to implement soon it will be possible to add a keymap for this without a dedicated builtin command so I would lean towards not merging this

cole-h commented:
Good to know, thanks! I'll close this and carry the patch locally until that comes about :)

cole-h commented:
Now that https://github.com/helix-editor/helix/pull/4709 has merged, here's a macro you can use to get this functionality!

```toml
[keys.normal.space]
"c" = "@:cd <C-r>%<C-w><ret>"
```

Pressing <kbd>space</kbd><kbd>c</kbd> will `cd` helix to the directory containing the current file!
